proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,316 @@
+
+Things we all agree about
+=========================
+
+Here are some points that everyone agrees about:
+
+1. ``base``, and other packages that come with GHC, should adhere rigorously to the PVP.
+
+2. Any complicated package, certainly including ``base``, has implementation "internals" that it may want to expose to hard-core users, but not to regular clients.
+   The right pattern for accommodating this is described in `Nikita's blog post <https://nikita-volkov.github.io/internal-convention-is-a-mistake/>`__: have two packages, an "internals" one exposing the internals, and a "stable" one that exposes the stable API. Both adhere to the PVP.
+
+3. We should use this pattern for ``base``.
+
+4. The Core Libraries Committee curates the API of ``base`` (here is `the charter <https://github.com/haskell/core-libraries-committee#base-package>`__), including:
+
+   - Types, specifically including what instances are exposed
+
+   - Semantics (including strictness)
+
+   - Performance
+
+   - Semantic changes to documentation (the charter says *"Documentation changes normally fall under GHC developers purview, except significant ones (e.g., adding or changing type class laws)."*)
+
+   Beyond that, it has no interest in the implementation details (e.g. alpha renaming, moving things between modules, comments).
+
+5. In curating the ``base`` API, it is immaterial where code lives.
+   For example, if a change to ``ghc-prim`` changes the ``base`` API (as defined above) the GHC developers must consult the CLC.
+   The fact that the change isn't physically part of the ``base`` package is immaterial.
+   (Incidentally, ``base`` and ``ghc-prim`` are both part of the same GitHub repository, which also includes GHC's source code.)
+
+Proposal
+========
+
+We propose to divide ``base`` into three packages:
+
+- ``ghc-internals``: exposes aspects of GHC's internals that may be of interest to "hard-core" developers interested in maximum performance (see `Nikita's blog post <https://nikita-volkov.github.io/internal-convention-is-a-mistake/>`__).
+  The API of ``ghc-internals`` is of no direct interest to the CLC --- only its effects on the API of base.
+
+- ``base``: as now, whose API is curated by CLC.
+  Depends on ``ghc-internals``, and hence on ``ghc-bignum`` and ``ghc-prim``.
+
+- ``ghc-experimental``, initially empty, depends on ``base`` and on ``ghc-internals``.
+  Functions and data types here are intended to have their ultimate home in base, but while they are settling down they are subject to much weaker stability guarantees.
+  Generally, new functions and types introduced in GHC Proposals would start their life here.
+  Example: new type families and type constructors for tuples, `GHC Proposal #475 <https://github.com/ghc-proposals/ghc-proposals/pull/475>`__.
+
+  Another example: future APIs to access RTS statistics, which are fairly stable and user-exposed, but which are (by design) coupled closely to GHC's runtime and hence may change.
```

Comment:
new para

As its name suggests, the API of `ghc-experimental` is curated by the GHC team, although the CLC is willing to offer (non-binding) opinions, if consulted.

Reply:
```suggestion
  Another example: future APIs to access RTS statistics, which are fairly stable and user-exposed, but which are (by design) coupled closely to GHC's runtime and hence may change.
  
  As its name suggests, the API of ``ghc-experimental`` is curated by the GHC team, although the CLC is willing to offer (non-binding) opinions, if consulted.
```